bcgov/healthgateway
# Update waitlist ticket check-in and release flow AB#15614

# Fixes or Implements [AB#15614](https://dev.azure.com/qslvic/304a1f8c-dace-4f85-adf3-bf563d5b3a39/_workitems/edit/15614)

## Description
1. Correctly destroy timeout's created to check-in the current ticket.

    - Reduce the number of `timeout()` calls to a "centralised" method to ensure that older timeouts are correctly destroyed and scheduled.

2. Add release procedures for logout, navigation and browser closure events.

    - Add `onbeforeunload` handler to dispatch release action
    - Registration and unregistration of the event is managed through Eventbus messsages.
    - On login, the handler is unregistered to prevent releasing the ticket while authenticating. Upon return via the redirectURL the listener will be reattached.

3. Update environment appsettings for the correct instance of tickets.healthgateway.gov.bc.ca.


## Testing

- [ ] Functional Tests Updated

## Notes
How to test

1. Ensure a clean slate (clear storage data for chrome users)
2. Load App
    - Note unload handler registration message in console.
3. Click login 
    - Note /Ticket POST
4. Sign In (any authentication channel)
    - On site return note the re-registration of unload handler
5. Navigate to multiple pages (this was one of the errors that produced multiple requests to /Ticket PUT
    - Note multiple timeouts are registered as the guard checks the validity of the ticket and schedules timeouts.
6. Wait for timeout to expire and trigger the non-breaking check-in request /Ticket PUT
7. Logout
    - Note that /Ticket Delete is called
8. Wait for the landing page to be presented after logging out.
9. Click on login 
    - Note the non-breaking request to /Ticket POST to get a new ticket.
 
To test the navigation (equivalent of closing the browser) at step 7 ensure preserve logs is enabled then navigate away to google (very little console and network noise), Note that /Ticket Delete is called.


## Items to Review:

-   [General PR Guidelines](https://github.com/bcgov/healthgateway/wiki/PRguidance)
